spring-cloud-gcp-docs/src/main/asciidoc/pubsub.adoc

```diff
@@ -51,39 +50,64 @@ In the example, we're setting the number of threads used by both the `Publisher`
 
 [source,yaml]
 ----
-spring.cloud.gcp.pubsub.subscriber.executorThreads=2
-spring.cloud.gcp.pubsub.publisher.executorThreads=2
+spring.cloud.gcp.pubsub.subscriber-executor-threads=2
+spring.cloud.gcp.pubsub.publisher-executor-threads=2
 ----
 
 The default value for both these properties is 4 threads.
 
 ==== Publishing to a topic
 
-`PubSubTemplate` provides an asynchronous method to publish messages to a Google Cloud Pub/Sub
-topic.
+`PubSubTemplate` provides asynchronous methods to publish messages to a Google Cloud Pub/Sub topic.
 
 [source,java]
 ----
-public ListenableFuture<String> send(final String topic, Message message)
+public ListenableFuture<String> publish(String topic, String payload, Map<String, String> headers)
 ----
 
-The `send()` method takes a Spring Message and it uses the built-in `SimpleMessageConverter` to
-convert from `Message` to the `PubsubMessage` accepted by the Google Cloud Java Library for Pub/Sub.
-`SimpleMessageConverter` copies the `Message` headers into `PubsubMessage` attributes and the
-`Message` payload into the `PubsubMessage` data.
-`SimpleMessageConverter` calls `toString()` in both the `Message` headers and payload, since
-`PubsubMessage` only supports `String` attributes and data at the transmission level.
-We're considering JSON object encoding/decoding in the future to support more object types.
+Publishes a message containing the `payload` and `headers` to the target `topic`.
+The `payload` is converted into a `byte[]` using `Charset.defaultCharset()`.
+
+[source,java]
+----
+public ListenableFuture<String> publish(String topic, String payload, Map<String, String> headers,
+Charset charset)
+----
+
+Publishes a message containing the `payload` and `headers` to the target `topic`.
+The `payload` is converted into a `byte[]` using the specified `charset`.
+
+[source,java]
+----
+public ListenableFuture<String> publish(String topic, byte[] payload, Map<String, String> headers)
+----
+
+Publishes a message containing the `payload` and `headers` to the target `topic`.
+
+[source,java]
+----
+public ListenableFuture<String> publish(String topic, ByteString payload, Map<String, String> headers)
+----
+
+Publishes a message containing the `payload` and `headers` to the target `topic`.
+The `payload` is in Google Cloud Pub/Sub Client Library's native payload format, `ByteString`.
+
+[source,java]
+----
+public ListenableFuture<String> publish(String topic, PubsubMessage pubsubMessage)
+----
+
+Low level method to publish a Google Cloud Pub/Sub `PubSubMessage` to the target `topic`.
```

Comment:
This whole section above just feels like a duplication of the javadoc. Can we just link to the javadoc instead of maintaining the same information in two places?

Reply:
We currently don't have Javadoc published anywhere (we should probably look into this, the API link from http://cloud.spring.io/spring-cloud-gcp/ should link there).

Instead of making this listing, I can just say there are `publish()` flavours that can handle different `Message` payload types such as `String`, `byte[]`, `ByteString` and `PubsubMessage`...

Reply:
Yes, this is better, and maybe create an issue for publishing javadocs.